encoder/vaapiencoder_vp9.cpp

```diff
@@ -0,0 +1,323 @@
+    }
+    else {
+        picParam->refresh_frame_flags = 0x01; //refresh last frame
+        picParam->pic_flags.bits.frame_type = kInterFrame;
+
+        ReferenceQueue::const_iterator it = m_reference.begin();
+
+        for (uint32_t i = 0; it != m_reference.end(); ++it, i++)
+            picParam->reference_frames[i] = (*it)->getID();
+
+        it = m_reference.begin();
+
+        // last/golden/alt is used as reference frame. L0 forward
+        picParam->ref_flags.bits.ref_frame_ctrl_l0 = 0x7;
+
+        // golden and alt are last KeyFrame
+        // last is last decoded frame
+        picParam->ref_flags.bits.ref_last_idx = 0;
+        picParam->ref_flags.bits.ref_gf_idx = 1;
+        picParam->ref_flags.bits.ref_arf_idx = 2;
+        picParam->pic_flags.bits.frame_context_idx = 0;
+    }
+
+    picParam->frame_width_src = width();
+    picParam->frame_height_src = height();
+    picParam->frame_width_dst = width();
+    picParam->frame_height_dst = height();
+
+    picParam->pic_flags.bits.show_frame = 1;
+
+    picParam->luma_ac_qindex = kDefaultQPValue;
+    picParam->luma_dc_qindex_delta = 1;
+    picParam->chroma_ac_qindex_delta = 1;
+    picParam->chroma_dc_qindex_delta = 1;
+
+    picParam->filter_level = kDefaultFilterLevel;
+    picParam->sharpness_level = kDefaultSharpnessLevel;
+
+    return true;
+}
+
+bool VaapiEncoderVP9::fill(
+    VAEncMiscParameterTypeVP9PerSegmantParam* segParam) const
+{
+    return true;
+}
+
+bool VaapiEncoderVP9::ensureSequence(const PicturePtr& picture)
+{
+    if (picture->m_type != VAAPI_PICTURE_I)
+        return true;
+
+    VAEncSequenceParameterBufferVP9* seqParam;
+    if (!picture->editSequence(seqParam) || !fill(seqParam)) {
+        ERROR("failed to create sequence parameter buffer (SPS)");
+        return false;
+    }
+    return true;
+}
+
+bool VaapiEncoderVP9::ensurePicture(const PicturePtr& picture,
+                                    const SurfacePtr& surface)
+{
+    VAEncPictureParameterBufferVP9* picParam;
+
+    if (!picture->editPicture(picParam) || !fill(picParam, picture, surface)) {
+        ERROR("failed to create picture parameter buffer (PPS)");
+        return false;
+    }
+    return true;
+}
+
+bool VaapiEncoderVP9::ensureQMatrix(const PicturePtr& picture)
+{
+    VAEncMiscParameterTypeVP9PerSegmantParam* segmentParam;
+
+    if (picture->m_type != VAAPI_PICTURE_I)
+        return true;
+
+    if (!picture->editQMatrix(segmentParam) || !fill(segmentParam)) {
+        ERROR("failed to create qMatrix");
+        return false;
+    }
+    return true;
+}
+
+bool VaapiEncoderVP9::referenceListUpdate(const PicturePtr& pic,
+                                          const SurfacePtr& recon)
+{
+
+    if (pic->m_type == VAAPI_PICTURE_I) {
+        m_reference.clear();
+        m_reference.insert(m_reference.end(), kMaxReferenceFrames, recon);
+    }
+    else {
+        m_reference.pop_front();
+        m_reference.push_front(recon);
```

Comment:
It is like this:

First key frame gets copied to reference list KKKKKKKK

For first Inter Frame (P frame in your comment) it will use the above and then update reference frames as

PKKKKKKK

Second inter frame (a.k.a. P frame)

PKKKKKKK

So it is updated at the front of the reference frames.

A run of this will look like:
...
[27117.381181][ctx 0x02000001]	reconstructed_frame = 0x04000005
[27117.381182][ctx 0x02000001]	reference_frames[0] = 0xffffffff
[27117.381183][ctx 0x02000001]	reference_frames[1] = 0xffffffff
[27117.381184][ctx 0x02000001]	reference_frames[2] = 0xffffffff
[27117.381185][ctx 0x02000001]	reference_frames[3] = 0xffffffff
[27117.381185][ctx 0x02000001]	reference_frames[4] = 0xffffffff
[27117.381186][ctx 0x02000001]	reference_frames[5] = 0xffffffff
[27117.381187][ctx 0x02000001]	reference_frames[6] = 0xffffffff
[27117.381187][ctx 0x02000001]	reference_frames[7] = 0xffffffff
...
[27117.396715][ctx 0x02000001]	reconstructed_frame = 0x04000006
[27117.396716][ctx 0x02000001]	reference_frames[0] = 0x04000005
[27117.396717][ctx 0x02000001]	reference_frames[1] = 0x04000005
[27117.396718][ctx 0x02000001]	reference_frames[2] = 0x04000005
[27117.396718][ctx 0x02000001]	reference_frames[3] = 0x04000005
[27117.396719][ctx 0x02000001]	reference_frames[4] = 0x04000005
[27117.396720][ctx 0x02000001]	reference_frames[5] = 0x04000005
[27117.396721][ctx 0x02000001]	reference_frames[6] = 0x04000005
[27117.396721][ctx 0x02000001]	reference_frames[7] = 0x04000005
...
[27117.411372][ctx 0x02000001]	reconstructed_frame = 0x04000007
[27117.411372][ctx 0x02000001]	reference_frames[0] = 0x04000006
[27117.411373][ctx 0x02000001]	reference_frames[1] = 0x04000005
[27117.411374][ctx 0x02000001]	reference_frames[2] = 0x04000005
[27117.411375][ctx 0x02000001]	reference_frames[3] = 0x04000005
[27117.411375][ctx 0x02000001]	reference_frames[4] = 0x04000005
[27117.411376][ctx 0x02000001]	reference_frames[5] = 0x04000005
[27117.411377][ctx 0x02000001]	reference_frames[6] = 0x04000005
[27117.411378][ctx 0x02000001]	reference_frames[7] = 0x04000005
...
[27117.424727][ctx 0x02000001]	reconstructed_frame = 0x04000008
[27117.424728][ctx 0x02000001]	reference_frames[0] = 0x04000007
[27117.424728][ctx 0x02000001]	reference_frames[1] = 0x04000005
[27117.424729][ctx 0x02000001]	reference_frames[2] = 0x04000005
[27117.424730][ctx 0x02000001]	reference_frames[3] = 0x04000005
[27117.424731][ctx 0x02000001]	reference_frames[4] = 0x04000005
[27117.424731][ctx 0x02000001]	reference_frames[5] = 0x04000005
[27117.424732][ctx 0x02000001]	reference_frames[6] = 0x04000005
[27117.424733][ctx 0x02000001]	reference_frames[7] = 0x04000005
....